{% from "macros/sequence-display.html" import render_sequence_modal %}

<h2>Homology to viral reference sequences</h2>

<p>
  This section shows the results from a BLAST homology search that was performed on a contig set generated using the de novo assembly tool SPADES.<br>
  The table below shows contigs that were at least 150 bp long and matched a known virus/viroid sequence, when performing a homology search against the core nucleotide database.<br>
  It can be sorted by alphabetical or numerical order by clicking on the column headers, and the number of entries shown can be adjusted using the dropdown menu above the table.<br>
  The results can be limited to a specific virus genus or species by typing its name in the search box.<br>
  The table can also be filtered to only show rows which meet different criteria (i.e. minimum coverage, terminology, and/or best contig per species).<br>
  A representative contig (the "best contig per species") is selected based on the preliminary blast homology search. Contigs assigned to the same species are ranked using a weighted indicator-based composite score that rewards several key features (% identity, bitscore, evalue, query coverage per subject, k-mer coverage of contig during assembly, contig length, alignment length, and annotation completeness). The contig with the highest score is then selected as the representative for that species.<br>
  The reads are also mapped back to the contigs to derive statistics on the number of reads mapped, mean depth, % coverage at 30X, and mean mapping quality. A confidence score is calculated based on these statistics to evaluate how well the raw sequencing data supports the assembled contigs.<br>
</p>

<p class="text-danger text-emphasis">
  Provide link to list of terms filtered in filterKeyWords.txt.
</p>
<p class="alert alert-warning">
  Erroneous sequences often result from sample contamination or from sequencing/assembly artefacts. It is the responsibility of the analyst to select the most appropriate sequence(s) for taxonomic identification based on the evidence shown. Furthermore, the results shown here <strong>cannot be used</strong> to make a confident assessment of the sample taxonomy. Taxonomic identity of a consensus sequence should be determined by an appropriate downstream analysis.
</p>

<ul class="lead">
  <li>
    {{ contigs_fasta | length }} contigs were assembled from quality-filtered reads assigned as viral or unclassified by Kraken.
  </li>
  <li>
    {{ blast_stats.count }}
    ({{ blast_stats.percent }}%)
    of contigs matched a virus or viroid sequence with BLAST.
  </li>
</ul>

<div class="my-3" style="overflow-x: auto;">

  {% if blast_hits | length %}
  <table id="blastTable" class="table table-lined tight text-center">
    <thead>
      <tr>
        {% for colname in blast_hits.columns_primary_display %}
        <th
          {% if blast_hits.COLUMN_METADATA[colname]['tooltip'] %}
          title="{{ blast_hits.COLUMN_METADATA[colname]['tooltip'] }}"
          data-bs-toggle="tooltip"
          data-bs-placement="top"
          {% endif %}
        >
          {{ blast_hits.COLUMN_METADATA[colname]['label'] }}
        </th>
        {% endfor %}
    </tr>
    <tr class="filters">
      {% for colname in blast_hits.columns_primary_display %}
        <th></th>
      {% endfor %}
    </tr>
  </thead>
    <tbody>
      {% for row in blast_hits %}
      {% set contig_id = row['qseqid'] %}
      <tr
        {% if row['bs_class'] %}
        class="alert-{{ row['bs_class'] }}"
        {% endif %}
      >
        {% for colname in blast_hits.columns_primary_display %}
        <td class="align-middle">
          {% if colname == 'evalue' %}
            <span style="white-space: nowrap;">{{ row[colname] }}</span>
          {% elif colname == 'contig_seq' %}
            {{ render_sequence_modal(contig_id, row[colname]) }}
          {% else %}
            {{ row[colname] }}
          {% endif %}
        </td>
        {% endfor %}
      </tr>
      {% endfor %}
    </tbody>
  </table>
  {% else %}
    {% if not blast_passed %}
    <p class="alert alert-danger">
      No viral sequences were identified with BLAST homology search.
    </p>
    {% else %}
    <p class="alert alert-info">
      No viral sequences were identified with BLAST homology search.
      <!-- TODO: verify this -->
    </p>
    {% endif %}
  {% endif %}
</div>

{% if blast_hits | length %}
<button
  type="button"
  class="btn btn-primary"
  data-bs-toggle="modal"
  data-bs-target="#consensusBlastModal"
>
  Consensus statistics
</button>

<a
  class="btn btn-primary hide-broken"
  href="{{ bam_html_file }}"
  target="_blank"
>
  Read alignment (BAM)
</a>

<button
  class="btn btn-primary"
  type="button"
  data-bs-toggle="modal"
  data-bs-target="#flagDefinitionsModal"
>
  Flag definitions
</button>
{% endif %}

{% include 'components/fasta-modal.html' %}
{% if blast_hits | length %}
  {% include 'components/blast-modal.html' %}
  {% include 'components/flag-definitions-modal.html' %}
{% endif %}

<script>
$(document).ready(function () {

  var table = $('#blastTable').DataTable({
    orderCellsTop: true,
  });

  var booleanColumns = [10, 11, 12];

  booleanColumns.forEach(function (colIdx) {
    var column = table.column(colIdx);
    var cell = $('#blastTable thead tr.filters th').eq(colIdx);

    var select = $('<select class="form-select form-select-sm">' +
                   '<option value="">All</option>' +
                   '<option value="TRUE">TRUE</option>' +
                   '<option value="FALSE">FALSE</option>' +
                   '</select>')
      .appendTo(cell)
      .on('change', function () {
        var val = $.fn.dataTable.util.escapeRegex($(this).val());
        column
          .search(val ? '^' + val + '$' : '', true, false)
          .draw();
      });
  });

});
</script>
