{% from "macros/sequence-display.html" import render_view_sequence_button %}
{% set term_filter_idx = blast_hits.columns_primary_display.index('term_filter') %}

<h2>Homology to viral reference sequences</h2>

<div class="alert alert-info">
  <div class="row align-items-center">
    <div class="col">
      The <strong>BLAST</strong> analysis compares the sequences assembled from your sample with a reference database of known viral sequences.

      The <strong>Terms to exclude</strong> column allows you to filter out sequences that match a 
      <a href="javascript:void(0);" data-bs-toggle="modal" data-bs-target="#filterKeywordsModal">
        pre-defined list of terms</a>.

      The <strong>Meets coverage</strong> column allows you to filter out sequences that are not well supported by the data.
    </div>
    <div class="col-auto">
      <button type="button" class="btn btn-primary btn-sm" data-bs-toggle="modal" data-bs-target="#blastInfoModal">
        About this analysis
      </button>
    </div>
  </div>
</div>

<ul class="lead">
  <li>
    {{ contigs_fasta | length }} contigs were assembled from quality-filtered reads assigned as viral or unclassified by Kraken.
  </li>
  <li>
    {{ blast_stats.count }}
    ({{ blast_stats.percent }}%)
    of contigs matched a virus or viroid sequence with BLAST.
  </li>
</ul>

<div class="my-3" style="overflow-x: auto;">

  {% if blast_hits | length %}
  <table id="blastTable" class="table table-lined tight text-center">
    <thead>
      <tr>
        {% for colname in blast_hits.columns_primary_display %}
        <th
          {% if blast_hits.COLUMN_METADATA[colname]['tooltip'] %}
          title="{{ blast_hits.COLUMN_METADATA[colname]['tooltip'] }}"
          data-bs-toggle="tooltip"
          data-bs-placement="top"
          {% endif %}
        >
          {{ blast_hits.COLUMN_METADATA[colname]['label'] }}
        </th>
        {% endfor %}
    </tr>
    <tr class="filters">
      {% for colname in blast_hits.columns_primary_display %}
        <th></th>
      {% endfor %}
    </tr>
  </thead>
    <tbody>
      {% for row in blast_hits %}
      {% set contig_id = row['qseqid'] %}
      <tr
        {% if row['bs_class'] %}
        class="table-{{ row['bs_class'] }} align-middle"
        {% endif %}
      >
        {% for colname in blast_hits.columns_primary_display %}
        <td class="align-middle">
          {% if colname == 'evalue' %}
            <span style="white-space: nowrap;">{{ row[colname] }}</span>
          {% elif colname == 'contig_seq' %}
            {{ render_view_sequence_button("Assembled sequence " + contig_id, row[colname]) }}
          {% elif colname == 'term_filter' %}
            {{ '' if row[colname] else 'Y' }}
          {% else %}
            {{ row[colname] }}
          {% endif %}
        </td>
        {% endfor %}
      </tr>
      {% endfor %}
    </tbody>
  </table>
  {% else %}
    {% if not blast_passed %}
    <p class="alert alert-danger">
      No viral sequences were identified with BLAST homology search.
    </p>
    {% else %}
    <p class="alert alert-info">
      No viral sequences were identified with BLAST homology search.
      <!-- TODO: verify this -->
    </p>
    {% endif %}
  {% endif %}
</div>

{% if blast_hits | length %}
<button
  type="button"
  class="btn btn-primary"
  data-bs-toggle="modal"
  data-bs-target="#consensusBlastModal"
>
  Contigs statistics
</button>

<a
  class="btn btn-primary hide-broken"
  href="{{ bam_html_file }}"
  target="_blank"
>
  Read alignment (BAM)
</a>

<button
  class="btn btn-primary"
  type="button"
  data-bs-toggle="modal"
  data-bs-target="#flagDefinitionsModal"
>
  Flag definitions
</button>
{% endif %}

<div class="modal fade" id="blastInfoModal" tabindex="-1" aria-labelledby="blastInfoModalLabel" aria-hidden="true">
  <div class="modal-dialog modal-lg">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title" id="blastInfoModalLabel">Homology search with BLAST</h5>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>

      <div class="modal-body">
        <ol>
          <li>
            Quality-filtered reads that were assigned as viral or unclassified by Kraken are <em>de novo</em> assembled into contigs (without a reference genome) with the <strong>SPAdes</strong> program. Contigs at least 150 bp long that match a known virus/viroid sequence are shown here, based on a BLAST homology search against the NCBI core_nt database.
          </li>
          <li>
            Sort the table by alphabetical or numerical order by clicking on column headers, and adjust the number of displayed entries using the dropdown menu above the table.
          </li>
          <li>
            The results can be limited to a specific virus genus or species by typing its name in the search box.
          </li>
          <li>
            The table can also be filtered to only show rows which meet different criteria (i.e. minimum coverage, taxon identified by BLAST does not match a pre-defined exclusion list, and/or best contig per species).
          </li>
          <li>
            The <strong>"Best contig per species"</strong> is selected based on the BLAST homology search results. Contigs assigned to the same species are ranked using a weighted indicator-based composite score that rewards several key features (% identity, bitscore, evalue, query coverage per subject, k-mer coverage of contig during assembly, contig length, alignment length, and annotation completeness). The contig with the highest score is then selected as the representative for that species.
          </li>
          <li>
            The reads are also mapped back to the contigs to derive statistics on the number of reads mapped, mean depth, % coverage at 30X, and mean mapping quality. A <strong>confidence score</strong> is calculated based on these statistics to evaluate how well the raw sequencing data supports the assembled contigs.
          </li>
        </ol>
      </div>

      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

{% if blast_hits | length %}
  {% include 'components/blast-modal.html' %}
  {% include 'components/flag-definitions-modal.html' %}
{% endif %}

<script>
$(document).ready(function () {

  var table = $('#blastTable').DataTable({
  orderCellsTop: true,
  order: [
    [18, 'desc'],
    [3, 'desc']
    ]
  });

  var booleanColumns = [11, 12];
  var termFilterIdx = {{ term_filter_idx }};

  booleanColumns.forEach(function (colIdx) {
    var column = table.column(colIdx);
    var cell = $('#blastTable thead tr.filters th').eq(colIdx);

    var select = $('<select class="form-select form-select-sm">' +
                   '<option value="">All</option>' +
                   '<option value="TRUE">TRUE</option>' +
                   '<option value="FALSE">FALSE</option>' +
                   '</select>')
      .appendTo(cell)
      .on('change', function () {
        var val = $.fn.dataTable.util.escapeRegex($(this).val());
        column
          .search(val ? '^' + val + '$' : '', true, false)
          .draw();
      });
  });
  (function () {
    var column = table.column(termFilterIdx);
    var cell = $('#blastTable thead tr.filters th').eq(termFilterIdx);

    $('<select class="form-select form-select-sm">' +
        '<option value="">All</option>' +
        '<option value="Y">Y</option>' +
        '<option value="blank">â€”</option>' +
      '</select>')
      .appendTo(cell)
      .on('change', function () {
        var val = $(this).val();
        if (val === "blank") {
          column.search('^$', true, false).draw();   // match empty cell
        } else {
          column.search(val ? '^' + val + '$' : '', true, false).draw();
        }
      });
  })();


});
</script>