{% from "macros/sequence-display.html" import render_sequence_modal %}

<h2>Mapping reads to viral references</h2>

<p class="text-end">
  <button type="button" class="btn btn-primary btn-sm" data-bs-toggle="modal" data-bs-target="#mappingInfoModal">
    About this analysis
  </button>
</p>

<div class="my-3" style="overflow-x: auto;">

  {% if mapping | length %}
  <table class="table table-lined tight text-center">
    <thead>
      <tr>
        {% for colname in mapping.columns_primary_display %}
        <th
          {% if mapping.COLUMN_METADATA[colname]['tooltip'] %}
          title="{{ mapping.COLUMN_METADATA[colname]['tooltip'] }}"
          data-bs-toggle="tooltip"
          data-bs-placement="top"
          {% endif %}
        >
          {{ mapping.COLUMN_METADATA[colname]['label'] }}
        </th>
        {% endfor %}
      </tr>
    </thead>
    <tbody>
      {% for row in mapping %}
      {% set accession = row[mapping.columns_primary_display[0]] %}
      <tr
        {% if row['bs_class'] %}
        class="alert-{{ row['bs_class'] }} align-middle"
        {% endif %}
      >
        {% for colname in mapping.columns_primary_display %}
        <td class="align-middle">
          {% if colname == 'evalue' %}
          <span style="white-space: nowrap;">{{ row[colname] }}</span>
          {% elif colname == 'consensus_seq' %}
          {{ render_sequence_modal(accession, row[colname]) }}
          {% else %}
          {{ row[colname] }}
          {% endif %}
        </td>
        {% endfor %}
      </tr>
      {% endfor %}
    </tbody>
  </table>
  {% else %}
  <p>
    No reference sequences available for mapping.
  </p>
  {% endif %}
</div>

<div class="modal fade" id="mappingInfoModal" tabindex="-1" aria-labelledby="mappingInfoModalLabel" aria-hidden="true">
  <div class="modal-dialog modal-lg">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title" id="mappingInfoModalLabel">Mapping reads to viral references</h5>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>

      <div class="modal-body">
        <ol>
          <li>
            Quality-filtered reads are mapped back to the viral reference sequences identified using the megablast homology search.
          </li>
          <li>
            References with <code>&gt;97%</code> similarity to a reference sequence are clustered using <strong>CD-HIT</strong> (only the longest reference sequence is retained).
          </li>
          <li>
            After mapping, a consensus sequence is generated with <strong>bcftools</strong> and <strong>bedtools</strong> for each reference sequence using the mapped reads. The consensus is generated by taking the most common nucleotide at each position in the reference sequence, based on the aligned reads.
          </li>
          <li>
            If there is a tie in the most common nucleotide, an ambiguity code is used to represent the possible nucleotides at that position. Any positions not covered by any reads are represented by "N" in the consensus sequence.
          </li>
          <li>
            <strong>Important note:</strong> the mapping-to-reference approach only works well when the reference sequence identified is closely related to the virus present in the sample. If the virus is highly divergent from known reference sequences, reads might not align well and it may not be possible to generate an accurate consensus sequence.
          </li>
          <li>
            The rows are sorted based on the <strong>recovered confidence score</strong>, which takes into consideration the number of reads mapped, mean depth, % coverage at 30X, and mean mapping quality, and is normalised to a value between 0 and 1. A higher score indicates higher confidence in the mapping result.
          </li>
        </ol>
      </div>

      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>
