{% from "macros/sequence-display.html" import render_sequence_modal %}

<h2>Mapping reads to viral references</h2>

<p>
  This section shows the results of mapping the quality-filtered reads back to the viral reference sequences identified using the megablast homology search.<br> 
  The references that show > 97% similarity are clustered using CD-hit (with the longest reference sequence being retained).<br>
  After mapping, a consensus sequence is generated with bcftools and bedtools for each reference sequence using the mapped reads. The consensuses are generated by taking the most common nucleotide at each position in the reference sequence, based on the aligned reads.<br>
  If there is a tie in the most common nucleotide, an ambiguity code is used to represent the possible nucleotides at that position. Any positions in the reference sequence that are not covered by any reads are represented by "N" in the consensus sequence.<br>
  Please note that the mapping back to reference approach only works well when the reference sequence identified is closely related to the virus present in the sample.<br>
  If the virus present in the sample is highly divergent from known reference sequences, then reads might not align well and it may not be possible to generate a consensus sequence that accurately represents the viral sequence in the sample.<br>
  The rows are sorted based on the recovered confidence score, which takes into consideration the number of reads mapped, mean depth, % coverage at 30X, and mean mapping quality, and is normalised to a value between 0 and 1. A higher score indicates higher confidence in the mapping result.<br>
</p>

<div class="my-3" style="overflow-x: auto;">

  {% if mapping | length %}
  <table class="table table-lined tight text-center">
    <thead>
      <tr>
        {% for colname in mapping.columns_primary_display %}
        <th
          {% if mapping.COLUMN_METADATA[colname]['tooltip'] %}
          title="{{ mapping.COLUMN_METADATA[colname]['tooltip'] }}"
          data-bs-toggle="tooltip"
          data-bs-placement="top"
          {% endif %}
        >
          {{ mapping.COLUMN_METADATA[colname]['label'] }}
        </th>
        {% endfor %}
      </tr>
    </thead>
    <tbody>
      {% for row in mapping %}
      {% set accession = row[mapping.columns_primary_display[0]] %}
      <tr
        {% if row['bs_class'] %}
        class="alert-{{ row['bs_class'] }} align-middle"
        {% endif %}
      >
        {% for colname in mapping.columns_primary_display %}
        <td class="align-middle">
          {% if colname == 'evalue' %}
          <span style="white-space: nowrap;">{{ row[colname] }}</span>
          {% elif colname == 'consensus_seq' %}
          {{ render_sequence_modal(accession, row[colname]) }}
          {% else %}
          {{ row[colname] }}
          {% endif %}
        </td>
        {% endfor %}
      </tr>
      {% endfor %}
    </tbody>
  </table>
  {% else %}
  <p>
    No reference sequences available for mapping.
  </p>
  {% endif %}
</div>
