<div class="row">
  <div class="col-auto">
    {% include "components/walltime.html" %}
  </div>
  <div class="col">
    <p class="small">
      This report describes a detailed analysis of your RNASeq data. A series of viral detection analyses were performed to evaluate potential viral material in your sample. Kraken and Kaiju show the many different taxa that your reads match to, while BLAST and "Reference mapping" show the alignment of your sequencing reads to known viral references. "Novel virus detection" shows an attempt to identify assembled sequences with viral characteristics.
      <a class="small" data-bs-toggle="modal" href="#overviewInfoModal" role="button">
        More info
      </a>
    </p>

    <p class="alert alert-warning small">
      <strong>Warning:</strong> Erroneous sequences often result from sample contamination or from sequencing/assembly artefacts. It is the responsibility of the analyst to select the most appropriate sequence(s) for taxonomic identification based on the evidence shown. Furthermore, the results shown here <strong>cannot be used</strong> to make a confident assessment of the sample taxonomy. Taxonomic identity of a consensus sequence should be determined by an appropriate downstream analysis.
    </p>

    <div class="modal fade" id="overviewInfoModal" tabindex="-1" aria-labelledby="overviewInfoModalLabel" aria-hidden="true">
      <div class="modal-dialog modal-lg modal-dialog-scrollable">
        <div class="modal-content">
          <div class="modal-header">
            <h5 class="modal-title" id="overviewInfoModalLabel">About this report</h5>
            <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
          </div>
          <div class="modal-body">
            <h6>Overview</h6>
            <p>
              A summary of the input parameters, quality control metrics and an overview of the viruses/viroids detected by both BLAST homology and direct read homology searches. Viruses reported here match at least one sequence assembled from your sample.
            </p>

            <h6>Kraken and Kaiju</h6>
            <p>
              These tools perform per-read taxonomic classification, in an effort to estimate the taxonomic distribution of your DNA sample. This provides an overview of the sample contents - often we expect to see a lot of host material and a small amount of viral material. It is normal to see hundreds of different taxa in these results, but you should pay attention to the distribution of read classification as this indicates contamination and complexity in your sample. A very clean sample might contain 98% host material, 1% viral material and 1% bacterial material.
            </p>

            <h6>BLAST</h6>
            <p>
              This analysis shows assembled DNA sequences that match viral reference sequences in a BLAST homology search. Reference sequence coverage is then derived by mapping sequencing reads back to these references.
            </p>

            <h6>Reference mapping</h6>
            <p>
              Shows consensus generated when mapping reads back to viral reference genomes detected by megablast (see the BLAST tab).
            </p>

            <h6>Novel virus detection</h6>
            <p>
              Attempts to reveal potential novel viral sequences derived by <em>de novo</em> assembly that have viral characteristics (based on <code>geNomad</code> and <code>hmmscan</code>), but did not match a known viral reference sequence.
            </p>
          </div>
          <div class="modal-footer">
            <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>

<div class="row my-5">
  <div class="col mr-5">
    {% include "components/inputs.html" %}
  </div>
  <div class="col">
    {% include "components/qc.html" %}
  </div>
</div>

{% include "components/summary.html" %}
